File: study/lua/src/yoka/extends/LabelExtend.lua
local Label = cc.Label
local scheduler=require("cocos.framework.scheduler")

function Label:updateTxtValue(num_value )

	if not num_value or num_value == "" then return end

	self._isPercent = false
	--百分比属性
	if type(num_value) =="string" and string.match(num_value, "%%") then
		num_value = string.sub(num_value, 1, -2)
		self._isPercent = true
	end

    local function txtRunAction(curr, dst)
	    self:stopUpdateSchedule()
	    -- self:scale(1)
	    -- self:doScaleAnimation()
	    
	    local step=30
	    self.stepValue=(dst-curr)/step
	    self.stepValue=self.stepValue>0 and math.ceil(self.stepValue) or math.floor(self.stepValue)
	    self.dstValue=dst
	    self.updateScheduler_=scheduler.scheduleGlobal(handler(self,self._updateScheduleFunc),1/30)
    end

    local value=tonumber(num_value)

    if(self.currValue==nil)then
    	local cur_value = self:getString()
    	if self._isPercent then
    		cur_value = string.sub(cur_value, 1, -2)
    	end
    	self.currValue= tonumber(cur_value)
    else
    	self.currValue=self.txt_temp_value
    end
    
   	if self.currValue~=value then
        txtRunAction(self.currValue,value)
    end

    self.txt_temp_value=value
    self:_formatNumber(num_value)
end

function Label:_formatNumber(num)
	local format=tostring(checknumber(num))
    self:setString(format..(self._isPercent and "%" or ""))
end

function Label:_updateScheduleFunc()
    if(self.currValue==nil)then return end
    self.currValue=self.currValue+self.stepValue
    if(self.currValue>=self.dstValue and self.stepValue>0)then
        self.currValue=self.dstValue
        self:stopUpdateSchedule()
    elseif(self.currValue<=self.dstValue and self.stepValue<0)then
        self.currValue=self.dstValue
        self:stopUpdateSchedule()
    end

    self:_formatNumber(self.currValue)
end

function Label:stopUpdateSchedule()
    if(self.updateScheduler_~=nil)then
        scheduler.unscheduleGlobal(self.updateScheduler_)
    end

    self:_formatNumber(self.currValue)
    self.updateScheduler_=nil
end
--数字增加 减少 动画效果 ---end


function Label:doScaleAnimation( duration )
	-- body
	local action1 = cc.ScaleTo:create(duration or 0.2,1.5)
	local action2 = cc.ScaleTo:create(duration or 0.2,1)
	local seq = cc.Sequence:create(action1,cc.DelayTime:create(0.1),action2)
	self:runAction(seq)
end

function Label:onExit( ... )
    -- body
    self:stopUpdateSchedule()
end

File: study/ts/script/extends/LabelExtend.ts
import { handler } from "../utils/handler";

// import { handler } from "../utils/handler";

export default class LabelExtend extends cc.Component{
    private _isPercent;
    private currValue;
    private txt_temp_value;
    private stepValue;
    private dstValue;
    updateTxtValue(num_value) {
        if (!num_value || num_value == '') {
            return;
        }
        this._isPercent = false;
        if (typeof(num_value) == 'string' && num_value.indexOf('%') >= 0) {
            num_value = num_value.substr(0, num_value.length-1);
            this._isPercent = true;
        }
        var value = parseInt(num_value);
        if (this.currValue == null) {
            var cur_value = this.label.string;
            if (this._isPercent) {
                cur_value = num_value.substr(0, num_value.length-1);
            }
            this.currValue = parseInt(cur_value);
        } else {
            this.currValue = this.txt_temp_value;
        }
        if (this.currValue != value) {
            this.txtRunAction(this.currValue, value);
        }
        this.txt_temp_value = value;
        this._formatNumber(num_value);
    }

    private _label;
    private get label(): cc.Label{
        if(!this._label) this._label = this.node.getComponent(cc.Label);
        return this._label;
    }

    txtRunAction(curr, dst) {
        this.stopUpdateSchedule();
        var step = 30;
        this.stepValue = (dst - curr) / step;
        this.stepValue = this.stepValue > 0 && Math.ceil(this.stepValue) || Math.floor(this.stepValue);
        this.dstValue = dst;
        this.schedule(this._updateScheduleFunc, 1 / 30);
    }

    _formatNumber(num) {
        var format = (parseInt(num)).toString();
        this.label.string = (format + (this._isPercent && '%' || ''));
    }

    _updateScheduleFunc() {
        if (this.currValue == null) {
            return;
        }
        this.currValue = this.currValue + this.stepValue;
        if (this.currValue >= this.dstValue && this.stepValue > 0) {
            this.currValue = this.dstValue;
            this.stopUpdateSchedule();
        } else if (this.currValue <= this.dstValue && this.stepValue < 0) {
            this.currValue = this.dstValue;
            this.stopUpdateSchedule();
        }
        this._formatNumber(this.currValue);
    }

    stopUpdateSchedule(){
        this.unschedule(this._updateScheduleFunc);
        this._formatNumber(this.currValue);
    }

    doScaleAnimation(duration) {
        var action1 = cc.scaleTo(duration || 0.2, 1.5);
        var action2 = cc.scaleTo(duration || 0.2, 1);
        var seq = cc.sequence(action1, cc.delayTime(0.1), action2);
        this.node.runAction(seq);
    }

    onDestroy(){
        this.unschedule(this._updateScheduleFunc);
    }

    onLoad(){
        this._label = this.node.getComponent(cc.Label);
    }

    getNumberValue(){
        return parseInt(this._label.string);
    }

    setNumberValue(value){
        this._label.string = value;
    }
}